### ISSUE TITLE
CA1801: ReviewUnusedParameters is reported but all parameters are used

### ISSUE THREAD

**Issue by thomaslevesque** [dotnet/roslyn-analyzers#1294]
#### Analyzer package

Example: Microsoft.CodeAnalysis.FxCopAnalyzers, v2.3.0-beta1

#### Analyzer

ReviewUnusedParameters

#### Repro steps

1. Write [this method](https://github.com/FakeItEasy/FakeItEasy/blob/b4083dbf31bec7d99b10f15198be4c71064cc79b/src/FakeItEasy.Analyzer/ArgumentConstraintNullabilityMismatchAnalyzer.cs#L81-L133):
```
        private static bool TryGetNameAndParameters(
            SyntaxNodeAnalysisContext context,
            SyntaxNode node,
            out string methodOrIndexerName,
            out IReadOnlyList<IParameterSymbol> parameters)
        {
            methodOrIndexerName = null;
            parameters = null;

            switch (node)
            {
                case InvocationExpressionSyntax invocation:
                {
                    var method = SymbolHelpers.GetCalledMethodSymbol(invocation, context);
                    if (method == null)
                    {
#if VISUAL_BASIC
                        var indexer = SymbolHelpers.GetAccessedIndexerSymbol(invocation, context);
                        if (indexer == null)
                        {
                            return false;
                        }

                        methodOrIndexerName = indexer.Name;
                        parameters = indexer.Parameters;
                        return true;
#else
                        return false;
#endif
                    }

                    parameters = method.Parameters;
                    methodOrIndexerName = method.GetDecoratedName();
                    return true;
                }
#if CSHARP
                case ElementAccessExpressionSyntax elementAccess:
                {
                    var indexer = SymbolHelpers.GetAccessedIndexerSymbol(elementAccess, context);
                    if (indexer == null)
                    {
                        return false;
                    }

                    methodOrIndexerName = indexer.Name;
                    parameters = indexer.Parameters;
                    return true;
                }
#endif
                default:
                    return false;
            }
        }
```
#### Expected behavior

No warning should be reported

#### Actual behavior

CA1801 is reported for `context` and `node`:

> Parameter context of method TryGetNameAndParameters is never used. Remove the parameter or use it in the method body.
> Parameter node of method TryGetNameAndParameters is never used. Remove the parameter or use it in the method body.

**Comment by onyxmaster** [comment]
Same here, I've got a much smaller repro (CA1801 for **q**):
```C#
void F(object q)
{
    switch (q)
    {
        case object o:
            o.GetHashCode();
            break;
    }
}
```

**Comment by zvirja** [comment]
@ivanbasov Could you please review this one? It's quite annoying issue preventing from painless usage of new C# features 😕

**Comment by jinujoseph** [comment]
cc @mavasani

**Comment by ivanbasov** [comment]
Thank you for reporting the issue! 
The issue has been recently fixed in IOperation. I have verified the fix with Microsoft.CodeAnalysis.FxCopAnalyzers, 2.6.0-beta1-62212-01

**Comment by zvirja** [comment]
@ivanbasov Thanks for the fix!

Do you also have any approx ETA when this package is published to Nuget?

**Comment by ivanbasov** [comment]
Thank you, @zvirja! We are going to ship new analyzers after shipping the new version of IOperation which is targeted for the end of 2017.

**Comment by mbarnet** [comment]
This issue appears on all code using Serialization, generally tagging the StreamingContext parameter which is required but not explicitly used in serialization code in classes using the interface. This has not been fixed as of v2.94 version of FxCop analytics.
